src/bsoncxx/decimal128.hpp

```diff
@@ -0,0 +1,100 @@
+// Copyright 2014 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <string>
+
+#include <bsoncxx/stdx/string_view.hpp>
+
+#include <bsoncxx/config/prelude.hpp>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+
+///
+/// Represents an IEEE 754-2008 Decimal128 value in a platform-independent way.
+///
+class BSONCXX_API decimal128 {
+   public:
+    ///
+    /// Constructs a Decimal128 value representing zero.
+    ///
+    decimal128() : _high(0), _low(0){};
+
+    ///
+    /// Constructs a Decimal128 from high and low 64-bit big-endian parts.
+    ///
+    /// @param high
+    ///   High bits
+    /// @param low
+    ///   Low bits
+    ///
+    decimal128(const uint64_t h, const uint64_t l) noexcept : _high(h), _low(l){};
+
+    ///
+    /// Constructs a Decimal128 from a string.
+    ///
+    /// @param str
+    ///   A string representation of a decimal number
+    ///
+    /// @throws bsoncxx::exception if the string isn't a valid Decimal128
+    /// representation.
+    ///
+    explicit decimal128(const stdx::string_view& str);
+
+    ///
+    /// Converts this decimal128 value to a string representation.
+    ///
+    /// @return A string representation of a IEEE 754-2008 decimal number.
+    ///
+    std::string to_string() const;
+
+    ///
+    /// @{
+    ///
+    /// Relational operators for decimal128
+    ///
+    /// @relates decimal128
+    ///
+    friend BSONCXX_API bool BSONCXX_CALL operator==(const decimal128& lhs, const decimal128& rhs);
+    friend BSONCXX_API bool BSONCXX_CALL operator!=(const decimal128& lhs, const decimal128& rhs);
+    ///
+    /// @}
+    ///
+
+    ///
+    /// Accessor for high 64 bits
+    ///
+    uint64_t high() const {
```

Comment:
Do you have a sense for what guidelines to use when deciding whether or not to inline methods in `BSONCXX_API`/`MONGOCXX_API` classes?  I'm not really sure yet.